### PR TITLE
Refactor NNCF compression statistics reporting API

### PR DESCRIPTION
**Pull request by andrey-churkin**

### CODE REVIEW

File: nncf/api/compression.py [hunk 1]
```diff
@@ -20,6 +21,28 @@
...
+
+        :return: A representation of the statistics as a human-readable string.
+        """
+
+    @abstractmethod
+    def as_dict(self) -> Dict[str, Any]:
+        """
+        Returns a representation of the statistics as built-in data types.
+
+        :return: A representation of the statistics as built-in data types.
```
Comment (vshampor) [inline nncf/api/compression.py:42]: Should add a note on what the string keys in the dict represent.

File: nncf/common/pruning/statistics.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        """
+        summary = {
+            'name': self.name,
+            'weight_shape': self.weight_shape,
+            'mask_shape': self.mask_shape,
+            'weight_pruning_level': self.weight_pruning_level,
+            'mask_pruning_level': self.mask_pruning_level,
+            'filter_pruning_level': self.filter_pruning_level,
+        }
+        return summary
```
Comment (vshampor) [inline nncf/common/pruning/statistics.py:63]: Note that this is essentially a copy-and-paste of the corresponding `self.__dict__`. If you decide to take this approach, then perhaps it would be easier to write a common `Statistics` function that iterates over `self.__dict__`, puts built-in types (int, double and str) into the retval dict as-is, puts the result of `.as_dict` for `Statistics` entries and ignores the rest (if encountered, spawn a debug message with a warning).

File: examples/classification/main.py [hunk 3]
```diff
@@ -453,7 +453,7 @@ def train_epoch(train_loader, model, criterion, criterion_fn, optimizer, compres
             config.tb.add_scalar("train/top1", top1.avg, i + global_step)
             config.tb.add_scalar("train/top5", top5.avg, i + global_step)
 
-            for stat_name, stat_value in compression_ctrl.statistics(quickly_collected_only=True).items():
+            for stat_name, stat_value in compression_ctrl.statistics(quickly_collected_only=True).as_dict().items():
```
Comment (vshampor) [inline examples/classification/main.py:456]: Can the `tb.add_scalar` handle nested dicts? Because your `as_dict` structure is, in general, nested.

**Comment by andrey-churkin** [comment]
Jenkins please retry a build

**Comment by andrey-churkin** [comment]
Jenkins please retry a build

File: nncf/api/composite_compression.py [hunk 4]
```diff
@@ -23,6 +24,38 @@
 ModelType = TypeVar('ModelType')
 
 
+class CompositeStatistics(Statistics):
```
Comment (alexsu52) [inline nncf/api/composite_compression.py:27]: If we look in terms of the main usage scenarios.  I guess, It is access to statistics values. We will see that the `CompressionAlgorithmController` returns `Statistics` object, but the `CompositeCompressionAlgorithmController` returns `CompositeStatistics` object. Currently, If the developer want to get sparsity_level, for example, then he should write the following code:

```
statistics = compression_ctrl.statistics()
if isinstance(statistics, CompositeStatistics):
    items = statistics.child_statistics
else:
    items = [statistics]

for item in items:
    if isinstance(item, (MagnitudeSparsityStatistics, RBSparsityStatistics, ConstSparsityStatistics)):
          sparsity_level = item.model_statistics.sparsity_level
    
```
before these changes it was like this: 

```
statistics = compression_ctrl.statistics()
sparsity_level = statistics['sparsity_level']
```

I was expecting the following:

```
statistics = compression_ctrl.statistics()
sparsity_level = statistics.magnitude_sparsity.sparsity_level
#or
magnitude_sparsity_statistics = statistics.get_statistics('magnitude_sparsity')
sparsity_level = magnitude_sparsity_statistics.sparsity_level
```
Can we make the statistics method return one type of object? Statistics container? @andrey-churkin, @vshampor Could you look once more at the proposal in this PR in terms of UX/DX?

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: beta/nncf/tensorflow/sparsity/rb/algorithm.py [hunk 5]
```diff
@@ -149,25 +151,26 @@ def raw_statistics(self):
...
-        })
-
         sparsity_levels = tf.keras.backend.batch_get_value(sparsity_levels)
         weights_percentages = [weights_number / total_weights_number * 100
                                for weights_number in weights_numbers]
         weights_percentages = tf.keras.backend.batch_get_value(weights_percentages)
         mask_sparsity = list(zip(mask_names, weights_shapes, sparsity_levels, weights_percentages))
-        raw_sparsity_statistics['sparsity_statistic_by_layer'] = []
+
+        # TODO(andrey-churkin): Why we use `mask_name` instead of `layer_name`?
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply beta/nncf/tensorflow/sparsity/rb/algorithm.py:160]: It's name from magnitude sparsity `raw_statistic` method. I used `wrapped_layer.name + '_rb_mask'`  name because it contains name of the layer but mask.name consist of `weight_attribute + '_mask'` thus less user friendly. mask_name not actually contains `mask.name`s but contains their user frendly names

File: beta/nncf/tensorflow/sparsity/rb/algorithm.py [hunk 6]
```diff
@@ -149,25 +151,26 @@ def raw_statistics(self):
...
+                SparsifiedLayerSummary(mask_name, weights_shape, sparsity_level, weights_percentage)
+            )
+
+        model_statistics = SparsifiedModelStatistics(sparsity_rate_for_model,
+                                                     sparsity_rate_for_sparsified_modules,
+                                                     sparsified_layers_summary)
+
+        target_level = self.loss.target_sparsity_rate
+        # TODO(andrey-churkin): Should be calculated when the distributed mode will be supported
+        masks_consistency = 1.0
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply beta/nncf/tensorflow/sparsity/rb/algorithm.py:173]: @vshampor Do we really want to report the `masks_consistency` statistic to users?

File: nncf/api/compression.py (hunk 1 above)
Comment (andrey-churkin) [reply nncf/api/compression.py]: The `as_dict()` method was removed after refactoring. For this reason, your comment is not applicable now.

File: beta/examples/tensorflow/classification/main.py [hunk 7]
```diff
@@ -221,7 +220,7 @@ def run(config):
             **validation_kwargs)
 
     logger.info('evaluation...')
-    print_statistics(compression_ctrl.statistics())
+    logger.info(compression_ctrl.statistics().as_str())
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply beta/examples/tensorflow/classification/main.py]: Done.

File: nncf/common/pruning/statistics.py (hunk 2 above)
Comment (andrey-churkin) [reply nncf/common/pruning/statistics.py]: The `as_dict()` method was removed after refactoring. I implemented a helper function that prepares statistics for the tensor board.

File: examples/classification/main.py (hunk 3 above)
Comment (andrey-churkin) [reply examples/classification/main.py]: The `tf.add_scalar` is not support nested dicts.

File: beta/nncf/tensorflow/sparsity/rb/algorithm.py [hunk 8]
```diff
@@ -149,25 +151,26 @@ def raw_statistics(self):
...
+
+        # TODO(andrey-churkin): Check that `mean_sparse_prob` is calculated correctly
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply beta/nncf/tensorflow/sparsity/rb/algorithm.py]: Fixed in ac19dc6.

File: beta/nncf/tensorflow/sparsity/rb/algorithm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply beta/nncf/tensorflow/sparsity/rb/algorithm.py]: Seems like a bug because `NNCFWrapper` can contain multiple operations with weights.

File: beta/nncf/tensorflow/pruning/base_algorithm.py [hunk 9]
```diff
@@ -305,23 +298,18 @@ def raw_statistics(self) -> Dict[str, object]:
...
         for mask_name, weights_shape, mask_shape, pruning_rate in mask_pruning:
-            raw_pruning_statistics['pruning_statistic_by_layer'].append({
-                'Name': mask_name,
-                'Weight\'s Shape': weights_shape,
-                'Mask Shape': mask_shape,
-                'PR': pruning_rate
-            })
-
-        return raw_pruning_statistics
+            # TODO(andrey-churkin): Should be calculated
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply beta/nncf/tensorflow/pruning/base_algorithm.py:306]: I created a ticket (55672) to track the progress of this task.

File: nncf/api/composite_compression.py (hunk 4 above)
Comment (andrey-churkin) [reply nncf/api/composite_compression.py]: Thank you very much for your suggestion. I fixed it in b5398b0.

**Comment by andrey-churkin** [comment]
Important changes:
- The `nncf/api/composite_compression.py` file was moved to `nncf/common/composite_compression.py`.
- The `statistics()` method was removed from `CompressionLoss` class.
- The `NNCFStatistics` class was introduced. All controllers should return an instance of this class from the `statistics()` method.

**Review by alexsu52 — APPROVED** [review]

**Review by asenina — APPROVED** [review]
👍

**Comment by andrey-churkin** [comment]
Jenkins please retry a build

**Comment by andrey-churkin** [comment]
@vshampor Could you please look at this PR?

File: beta/nncf/tensorflow/sparsity/rb/algorithm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply beta/nncf/tensorflow/sparsity/rb/algorithm.py:173]: @vshampor Could you please answer the question above?

[the remark this answers is not in the document]
Comment (vshampor) [reply beta/nncf/tensorflow/sparsity/rb/algorithm.py:173]: I have had no idea what mask consistency is before you had drawn my attention to it, and it is not immediately obvious to me what this parameter means in the context of sparsity and how it impacts the sparsity outputs. I think that we can omit reporting it for now.
